### ISSUE TITLE
twisted.web.static.File can cause open-file starvation from not closing files after HEAD requests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7846]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7846|
|Type|defect|
|Created|2015-04-14 05:14:56Z|
|Branch|https://github.com/twisted/twisted/tree/fd-starvation-7846-2|

<details><summary>Searchable metadata</summary>

```
trac-id__7846 7846
type__defect defect
reporter__hawkowl hawkowl
priority__high high
milestone__ 
branch__branches_fd_starvation_7846_2 branches/fd-starvation-7846-2
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1428988496625585 1428988496625585
changetime__1429858943691419 1429858943691419
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44397]) Branching to fd-starvation-7846.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl removed owner|
|-|-|

Patch applied, buildbots spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes looks good. Many thanks for the bugfix.

-----

here are just minor comments.

To prevent future errors of the same type maybe we should enclose all code using the opened file in a method and wrap it in a try/finally block

I find it hard to read the new tests and comments are not pep8 compatible.

Just like you set `request.method = 'HEAD'` it would help to also see `request.method = 'GET'` explicitly set in the test for GET...

Since the new tests are in fact testing the render method, maybe they should start with test_render*.

The MockFile is duplicated across the tests... but why not use StringIO?

Do we care about file's content in the context of this test? If not, I think that `path.setContent("foo")` could be left out.

Also, I see duplicate code in arranging the mocked FilePath resource for this test. Maybe it can be extracted into a separate method.

Does this test need a reactor? If not maybe you can use `successResultOf` and `failureResultOf`

I think that we should write test using the same attention as main code and not include unnecessary lines.


Please review my comments and let me know what do you think.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [44484]) Branching to fd-starvation-7846-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl removed owner|
|-|-|

Hi,

Since the t.web.static porting got through -- I ported this to trunk.

I cleaned up the tests, let me know what you think.

-hawkie

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Many thanks for the changes!

Even if this code is already in trunk, as wsanchez also mentioned, we should wither guard the opened file with a context for try/finally

I think that there is a great change that in the future someone else to return from render_GET from another place and forgetting to close the file.

--------

The tests are much better... still not AAA :)

StringIO already provides the closed attribute why not reuse StringIO?

For the test checking the GET method I don't see any explicit call to `request.method = b'GET'` since GET part is really important for this test I think that it should be explicitly set.

You don't really need a new test case for that. You can use a factory helper to create the mocked FilePath resource and its fake file object

```

resource, requestedFile = self.makeFilePathWithRiggedOpen()

```


where makeFilePathWithRiggedOpen is similar to the setup code but returns the 2 new objects rather than setting them as instance members.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Thanks for the review, Adi.

- I think wsanchez was talking about the tests (which I changed), not the implementation. Because of the nature of producers, you can't put the open in a with -- this bug was caused by stuff not being done in the producer.

- I didn't realise StringIO would do this... I'll make the change.

- The other parts might make it nicer to understand. I'll see how they go.

-Hawkie

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl removed owner|
|-|-|

I've made some changes. Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks.

Since the fix already hit the trunk not sure if this is still a bugfix.

I think that we should rename the ticket to mention that this is a refactorization and make the news fragment a .misc

In twisted/web/static.py the new comments are good but they should be proper sentences, with full stop.

The commend for cache response should be update as the setLastModified() is a messed up method. It does not only set last modified but also set the response code. I had to dig a bit in the code to understand how it works.

Instead of getmtime (since we touched this line) I think that we should update it to getModificationTime

As much as possible I try to write code which don't requires monkey pathing in tests but rather tries to arrange the test using public api. Tests with monkey paths are fragile.

In the case of cached get, I think that test can be arranged by setting the if-modified-since header on the request to that setLastModified() will return http.CACHE without mocking the method... we can mock the clock if needed.

Would but nice to have tests in AAA.

Tests call successResultOf but they don't check the result value.


Sorry for the pedantic review. If you are not in the mood for changes feel free to land with current tests as they are better than current trunk.

I ended up with this code but I don't know if it is any better.

```

        try:
            fileForReading = self.openForReading()
        except IOError as e:
            import errno
            if e.errno == errno.EACCES:
                return self.forbidden.render(request)
            else:
                raise

        if not self._canContinue(request):
            fileForReading.close()
            return b''

        producer = self.makeProducer(request, fileForReading)
        producer.start()

        # Make sure the connection doesn't get closed.
        return server.NOT_DONE_YET

    def _canContine(self, request):
        if request.setLastModified(self.getmtime()) is http.CACHED:
            # setLastModified will also set the response code.
            return False

        if request.method == b'HEAD':
            self._setContentHeaders(request)
            return False

        return True

```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

also, instead of self.assertEqual(fakeFile.closed, True) I think that you can use self.assertTrue(fakeFile.closed)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44539]) Merge fd-starvation-7846-2: Tests for twisted.web.static.File FD starvation

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7846](https://github.com/twisted/twisted/issues/7846)
